File: python_problems/combinatorics/combination_sum.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result: List[List[int]] = []
        path: List[int] = []

        def dfs(start: int, path_sum: int):
            if path_sum == target:
                result.append(path.copy())
                return

            for i in range(start, len(candidates)):
                val = candidates[i]
                if path_sum + val > target:
                    return

                path.append(val)
                dfs(i, path_sum + val)
                path.pop()

        dfs(0, 0)
        return result

File: practice/src/main/java/com/hegde/practice/combinatorics/CombinationSum.java
package com.hegde.practice.combinatorics;

import java.util.ArrayList;
import java.util.List;

/**
 * https://leetcode.com/problems/combination-sum/
 */
public class CombinationSum {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(candidates, 0, target, new ArrayList<>(), result);
        return result;
    }

    private void dfs(int[] candidates, int start, int target, List<Integer> path,  List<List<Integer>> result){
        if(target == 0){
            result.add(new ArrayList<>(path));
            return;
        }
        if(target < 0)
            return;
        for(int i = start; i < candidates.length; i++){
            path.add(candidates[i]);
            dfs(candidates, i, target - candidates[i], path, result);
            path.removeLast();
        }
    }
}
